topology/probes/libvirt/libvirt.go

```diff
@@ -0,0 +1,267 @@
+		logging.GetLogger().Error("XML parsing error: %s", err)
+		return
+	}
+	host, _ := domain.GetName()
+	for i := range d.Interfaces {
+		itf := &d.Interfaces[i]
+		if constraint == "" || constraint == itf.Alias.Name {
+			itf.Host = host
+			*bag = append(*bag, itf)
+		}
+	}
+}
+
+// registerInterfaces puts the information collected in the graph
+// interfaces is an array of collected information.
+func (probe *Probe) registerInterfaces(interfaces []*Interface) {
+	probe.graph.Lock()
+	defer probe.graph.Unlock()
+	for _, itf := range interfaces {
+		name := itf.Target.Device
+		if name == "" {
+			continue
+		}
+		probe.interfaceMap[name] = itf
+		logging.GetLogger().Debugf(
+			"Libvirt interface %s on %s", name, itf.Host)
+		node := probe.graph.LookupFirstNode(
+			graph.Metadata{"Type": "tun", "Name": name})
+		if node != nil {
+			probe.enrichNode(node, itf)
+		}
+	}
+}
+
+// enrichNode adds the libvirt interface information to a node of the graph
+func (probe *Probe) enrichNode(node *graph.Node, itf *Interface) {
+	logging.GetLogger().Debugf("enrich %s", itf.Alias.Name)
+	tr := probe.graph.StartMetadataTransaction(node)
+	tr.AddMetadata("Libvirt.MAC", itf.Mac.Address)
+	tr.AddMetadata("Libvirt.Domain", itf.Host)
+	address := itf.Address
+	formatted := fmt.Sprintf(
+		"%s:%s.%s.%s.%s", address.Type, address.Domain, address.Bus,
+		address.Slot, address.Function)
+	tr.AddMetadata("Libvirt.Address", formatted)
+	tr.AddMetadata("Libvirt.Alias", itf.Alias.Name)
+	tr.AddMetadata("PeerIntfMAC", itf.Mac.Address)
+	tr.Commit()
+}
+
+// Start get all domains attached to a libvirt connection
+func (probe *Probe) Start() {
+	// The event loop must be registered WITH its poll loop active BEFORE the
+	// connection is opened. Otherwise it just does not work.
+	if err := libvirtgo.EventRegisterDefaultImpl(); err != nil {
+		logging.GetLogger().Errorf("libvirt event handler:  %s", err)
+	}
+
+	go func() {
+		for {
+			if err := libvirtgo.EventRunDefaultImpl(); err != nil {
+				logging.GetLogger().Errorf("libvirt poll loop problem: %s", err)
+			}
+		}
+	}()
+
+	conn, err := libvirtgo.NewConnectReadOnly(probe.uri)
+	if err != nil {
+		logging.GetLogger().Errorf("Failed to create libvirt connect")
+		return
+	}
+	probe.conn = conn
+	callback := func(
+		c *libvirtgo.Connect, d *libvirtgo.Domain,
+		event *libvirtgo.DomainEventLifecycle,
+	) {
+		if event.Event == libvirtgo.DOMAIN_EVENT_STARTED {
+			var interfaces []*Interface
+			probe.queryDomain(d, &interfaces, "")
+			probe.registerInterfaces(interfaces)
+		}
+	}
+	probe.cidLifecycle, err = conn.DomainEventLifecycleRegister(nil, callback)
+	if err != nil {
+		logging.GetLogger().Errorf(
+			"Could not register the lifecycle event handler %s", err)
+	}
+	callbackDeviceAdded := func(
+		c *libvirtgo.Connect, d *libvirtgo.Domain,
+		event *libvirtgo.DomainEventDeviceAdded,
+	) {
+		var interfaces []*Interface
+		probe.queryDomain(d, &interfaces, event.DevAlias)
```

Comment:
this logic may be moved also here
https://github.com/skydive-project/skydive/pull/1424